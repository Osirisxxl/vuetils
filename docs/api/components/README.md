# Vue components

Creating components could be verbose sometimes. Those are usefull method for dealing with meta patterns on component creation.

!!!include(docs/api/methods/bemModifiers.md)!!!
