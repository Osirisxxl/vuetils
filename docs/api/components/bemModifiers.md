## bemModifiers

Returns `Array` with class generator for vue with [BEM Semantic](http://getbem.com/introduction/)

- **Returns:** `[ bockOrElementDefiniton, { ...modifiers: Boolean }]`

- **Example:**

```js
Vue.use(Vuetils);

export default {
  name: "DumbComponent",
  data() {
    return {
        visible: true,
        primary: false,
    }
  }
};
```

```html
<template>
    <div :class="$bemModifiers('block', { visible })">
        <h1 :class="$bemModifiers('block__element', { primary })"></h1>
    </div>
</template>
```