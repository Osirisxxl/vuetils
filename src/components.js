const toKebabCase = string =>
  string
    .replace(/([a-z])([A-Z])/g, "$1-$2")
    .replace(/\s+/g, "-")
    .toLowerCase();

export const bemModifiers = function(blockOrElementName, modifiers = {}) {
  return [
    blockOrElementName,
    Object.entries(modifiers).reduce(
      (modifiers, [name, value]) => ({
        ...modifiers,
        [`${blockOrElementName}--${toKebabCase(name)}`]: !!value
      }),
      {}
    )
  ];
};
