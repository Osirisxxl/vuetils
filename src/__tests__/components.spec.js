import { shallowMount, createLocalVue } from "@vue/test-utils";
import { bemModifiers } from "../components";
import Vuetils from "../index";

const localVue = createLocalVue();

localVue.use(Vuetils);

describe("bemModifiers", () => {
  it("should generate class with empty args", () => {
    expect(bemModifiers("block__element")).toEqual(["block__element", {}]);
  });

  it("should generate modifiers", () => {
    expect(
      bemModifiers("block__element", {
        visible: true,
        touchable: 1,
        isGhost: "yes",
        primary: null
      })
    ).toEqual([
      "block__element",
      {
        "block__element--visible": true,
        "block__element--touchable": true,
        "block__element--is-ghost": true,
        "block__element--primary": false
      }
    ]);
  });

  it("should be defined as prototype function", () => {
    const wrapper = shallowMount(
      {
        name: "DumbComponent",
        template: `
                <div :class="$bemModifiers('block', { visible })">
                    <h1 :class="$bemModifiers('block__element', { primary })"></h1>
                </div>
            `,
        data() {
          return {
            visible: true,
            primary: false
          };
        }
      },
      { localVue }
    );

    expect(wrapper.html()).toMatchSnapshot();
  });
});
