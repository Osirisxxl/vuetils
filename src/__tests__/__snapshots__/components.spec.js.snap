// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bemModifiers should be define as prototype function 1`] = `
<div class="block block--visible">
  <h1 class="block__element"></h1>
</div>
`;
